### PR TITLE
Migrate `Play-Run-Support` to pure Java implementation

### PR DESCRIPTION
**Pull request by ihostage**
**As a preparation for:**
- #11990

**Depends on:**
- https://github.com/playframework/play-file-watch/pull/242

These changes are only refactoring code and rewriting by pure Java and nothing changing in logic.
Potential changes in logic of `play-run-support` will be a part of the other pull requests are related with supporting Maven/Gradle.

### CODE REVIEW

**Comment by mkurz** [comment]
I rebased on main, I hope you don't mind ;)

**Comment by mkurz** [comment]
So I see these are just the changes take from #12006 from `dev-mode/play-run-support/` (?)

**Comment by ihostage** [comment]
> I rebased on main, I hope you don't mind ;)

Nop, I don't, of course 😉 

> So I see these are just the changes take from #12006 from `dev-mode/play-run-support/` (?)

Yes, they are a 99% of changes this folder from `gradle` branch. But what is important, these changes don't have a logically changes, only language migration. And if we need a change a logic to support both build system I will make these changes in separated PR.

**Comment by ihostage** [comment]
FYI @mkurz I rebased this branch into `main`

File: dev-mode/sbt-plugin/src/main/scala/sbt/PlayRun.scala [hunk 1]
```diff
@@ -188,20 +195,26 @@ object PlayRun {
...
-      watched.watchSources(state).iterator.flatMap(new PlaySource(_).getPaths).collect {
-        case p if Files.exists(p) => better.files.File(p)
-      }
+    val sourcesFinder: Supplier[java.lang.Iterable[java.io.File]] = () => {
+      watched
+        .watchSources(state)
+        .iterator
+        .flatMap(new PlaySource(_).getPaths)
+        .collect {
+          case p if Files.exists(p) => better.files.File(p).toJava
```
Comment (mkurz) [inline dev-mode/sbt-plugin/src/main/scala/sbt/PlayRun.scala:204]:
```suggestion
          case p if Files.exists(p) => p.toFile
```
Now better-files is not necessary anymore and the dependency can be removed also

**Comment by mkurz** [comment]
@ihostage I rebased this branch onto main, there were _no_ conflicts. Hmm.. maybe I made a mistake and didn't force pull before... Because now this commit and the first commit of the gradle branch differ now: faad519974ecd0a6b7b33a3aca83dc9fc9d27687..b32a8482ba9e92acd119131aabb4368e13a728e5
Can you please correctly force push this branch to it's correct sha-id?
I don't know what is wrong, I think this branch actually point to b32a8482ba9e92acd119131aabb4368e13a728e5?
Can you please check and push? Thanks!

**Comment by ihostage** [comment]
> If you can correct this I would be happy 😉

Of course, no problem. I updated a `play-file-watch` version and force re-pushed my version of branch 😉

**Comment by ihostage** [comment]
Plus cherry-pick your commit [502668bd70269897365a78ee55e10dcbd5fcf342](https://github.com/playframework/playframework/commit/502668bd70269897365a78ee55e10dcbd5fcf342)

File: dev-mode/sbt-plugin/src/test/scala/play/sbt/PlayRunHookSpec.scala [hunk 2]
```diff
@@ -29,13 +33,15 @@ class PlayRunHookSpec extends Specification {
         override def beforeStarted(): Unit = throw new HookMockException()
       }
 
-      hooks.run(_.beforeStarted()) must throwA[HookMockException]
+      RunHooksRunner.run(hooks.asJava, _.beforeStarted()) must throwA[RuntimeException].like {
+        case re => re.getCause must haveClass[HookMockException]
+      }
```
Comment (ihostage) [inline dev-mode/sbt-plugin/src/test/scala/play/sbt/PlayRunHookSpec.scala:38]: Small changes in the behaviour of `RunHooksRunner`. But I don't think that these changes is real matter. Whatever it was, the exception will be thrown and printed in the console with his cause 🤷‍♂️

**Comment by ihostage** [comment]
Finally, it is green now 🤪

File: dev-mode/play-run-support/src/main/java/play/runsupport/DevServerRunner.java [hunk 3]
```diff
@@ -43,7 +43,7 @@ private ReloadableServer getReloadableServer(
       server = (ReloadableServer) m.invoke(null, reloader, httpPort, httpAddress);
     } else {
       var m = mainClass.getMethod("mainDevOnlyHttpsMode", BuildLink.class, int.class, String.class);
-      server = (ReloadableServer) m.invoke(null, reloader, httpPort, httpAddress);
+      server = (ReloadableServer) m.invoke(null, reloader, httpsPort, httpAddress);
```
Comment (ihostage) [inline dev-mode/play-run-support/src/main/java/play/runsupport/DevServerRunner.java:46]: Damn 🤦‍♂️ Thanks, Matthias! 🙏

File: dev-mode/play-run-support/src/main/java/play/runsupport/DevServerReloader.java [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+import play.core.BuildLink;
+import play.dev.filewatch.FileWatchService;
+import play.dev.filewatch.FileWatcher;
+import play.runsupport.CompileResult.CompileFailure;
+import play.runsupport.CompileResult.CompileSuccess;
+import play.runsupport.classloader.DelegatedResourcesClassLoader;
+
+class DevServerReloader implements BuildLink, Closeable {
```
[the remark this answers is not in the document]
Comment (ihostage) [reply dev-mode/play-run-support/src/main/java/play/runsupport/DevServerReloader.java:32]: Yes. Visibility of some classes and this module in general are a question for future discussions and improvements. But not today, maybe in the future 😄

File: dev-mode/play-run-support/src/main/java/play/runsupport/classloader/DelegatedResourcesClassLoader.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import java.net.URL;
+import java.util.Enumeration;
+import java.util.Objects;
+
+/** A ClassLoader that only uses resources from its parent */
+public class DelegatedResourcesClassLoader extends NamedURLClassLoader {
+
+  public DelegatedResourcesClassLoader(String name, URL[] urls, ClassLoader parent) {
+    super(name, urls, parent);
+    Objects.requireNonNull(parent);
```
[the remark this answers is not in the document]
Comment (ihostage) [reply dev-mode/play-run-support/src/main/java/play/runsupport/classloader/DelegatedResourcesClassLoader.java:17]: Yep 😂

File: dev-mode/sbt-plugin/src/main/scala/play/sbt/run/package.scala [hunk 6]
```diff
@@ -12,13 +14,13 @@ package object run {
...
-    def error(message: => String): Unit   = in.error(message)
-    def trace(t: => Throwable): Unit      = in.trace(t)
-    def success(message: => String): Unit = in.success(message)
+    def verbose(message: Supplier[String]): Unit = in.verbose(message.get())
+    def debug(message: Supplier[String]): Unit   = in.debug(message)
+    def info(message: Supplier[String]): Unit    = in.info(message)
+    def warn(message: Supplier[String]): Unit    = in.warn(message)
+    def error(message: Supplier[String]): Unit   = in.error(message)
+    def trace(t: Supplier[Throwable]): Unit      = in.trace(t)
+    def success(message: Supplier[String]): Unit = in.success(message.get())
```
[the remark this answers is not in the document]
Comment (ihostage) [reply dev-mode/sbt-plugin/src/main/scala/play/sbt/run/package.scala:23]: Thanks, Matthias! 👍 
Yes, I saw it on your personal activity dashboard on GitHub 😉 Time to time I check it to see what you are also doing and find something interesting for itself 😊

**Comment by ihostage** [comment]
> I very carefully reviewed each line... And also tested a bit locally. Also thanks to our many scripted tests I feel quite comfortable this pull request should be ok now.

It was great, thank you a lot! 👍 

> Please take a look at my commits I pushed, is there anything you disagree with? If so please let me know.

I checked all of them immediately after you had pushed 😄 All changes are ok to me 🤝 
Also, I think our works made a sources of `play-run-support` more clear and better 😊 

> Also please look at the last two commits, where I fixed the initialization of the reloader because in original scala code these were lazy vars, so they should come after the "before started" hooks, otherwise file watcher would already run (and the `assetsClassLoader` function would be processed, which could be anything set by the user).

Yes, I saw it. It's my fault while migrating from lazy vars, sorry 😞

**Pull request merged**